.github/workflows/upload_binary.yml

```diff
@@ -40,8 +40,11 @@ jobs:
           python -m pip install pyinstaller
 
       - name: Build binary
-        run: |
-          python -m PyInstaller -F --name ${{ matrix.asset_name }} --add-data 'src/blib2to3${{ matrix.pathsep }}blib2to3' src/black/__main__.py
+        run: >
+          python -m PyInstaller -F --name ${{ matrix.asset_name }} --add-data
+          'src/blib2to3${{ matrix.pathsep }}blib2to3' --hidden-import platformdirs.unix
+          --hidden-import platformdirs.macos --hidden-import platformdirs.windows
+          src/black/__main__.py
```

Comment:
Just a question - Is there a fix we can do to platformdirs to not require this? I guess a follow on question is what is the difference with what platformdirs does vs. appdirs? Does it divide up each platforms into sub modules?

Reply:
It looks like `appdirs` [doesn't use dynamically loaded modules per platform](https://github.com/ActiveState/appdirs/blob/master/appdirs.py), while `platformdirs` [does](https://github.com/platformdirs/platformdirs/blob/main/src/platformdirs/__init__.py#L18-L28).

Reply:
The fact platformdirs breaks PyInstaller doesn't surprise me knowing how it works, but what does is that we have conditional imports in our code and yet they're not broken? The main culprit I have in mind is `typing-extensions`... it's not like the base install of Black has dependencies that then unconditionally depend on it. I guess sys.version_info guarded imports are treated correctly? 

Or I could be misunderstanding all of this which is totally possible as it's 11 PM over here ^^

Reply:
According to PyInstaller:

> Hidden imports can occur when the code is using __import__(), importlib.import_module() or perhaps exec() or eval(). Hidden imports can also occur when an extension module uses the Python/C API to do an import. When this occurs, Analysis can detect nothing. There will be no warnings, only an ImportError at run-time.

That seems to suggest that conditional imports are fine, but dynamic imports aren't. I haven't looked into exactly how PyInstaller detects imports, but I suspect they some version of searching for import statements.

Reply:
Oh right, conditional import != hidden import. G'ah I should really get some sleep :sweat_smile: